src/lib/optimizer/strategy/constant_calculation_rule.cpp

```diff
@@ -47,54 +62,107 @@ bool ConstantCalculationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& n
       resolve_data_type(*expression_type, [&](auto type) { value = _calculate_expression(type, expression); });
     }
 
-    // If the value is std::nullopt, then the expression could not be resolved at this point,
-    // e.g. because it is not an arithmetic operator.
-    if (value != std::nullopt &&
-        _replace_expression_in_parents(node, node->output_column_references()[column_id], *value)) {
-      // If we successfully replaced the occurrences of the expression in the parent tree,
-      // we can remove the column here.
-      _remove_column_from_projection(projection_node, column_id);
+    if (value != std::nullopt) {
+      column_reference_to_value_map.emplace(projection_node->output_column_references()[column_id], *value);
     }
   }
-
-  return _apply_to_children(node);
 }
 
-bool ConstantCalculationRule::_replace_expression_in_parents(const std::shared_ptr<AbstractLQPNode>& node,
-                                                             const LQPColumnReference& expression_column,
-                                                             const AllTypeVariant& value) {
-  auto parent_tree_changed = false;
-  for (auto parent : node->parents()) {
-    if (parent->type() != LQPNodeType::Predicate) {
-      parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
-      continue;
-    }
+bool ConstantCalculationRule::_replace_column_references_in_tree(
```

Comment:
Now that I think about it: You're just replacing column references in Projections and Predicates. What about those in all the other node types? Those should be covered as well, and the rule should also detect whether a node type isn't covered, e.g. because it was newly introduced, and Fail()/handle this gracefully.